// @mon-foyer/server/dist/utils/mailer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.sendInviteEmail = sendInviteEmail;
async function sendInviteEmail(to, name, inviteUrl) {
    const apiKey = process.env.RESEND_API_KEY;
    if (!apiKey) {
        console.warn('[mailer] RESEND_API_KEY not set — skipping email');
        return false;
    }
    const from = process.env.RESEND_FROM ?? 'onboarding@resend.dev';
    let res;
    try {
        res = await fetch('https://api.resend.com/emails', {
            method: 'POST',
            headers: {
                Authorization: `Bearer ${apiKey}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                from,
                to: [to],
                subject: 'Invitation à rejoindre Mon Foyer 🏠',
                html: `
          <div style="font-family:sans-serif;max-width:480px;margin:auto">
            <h2>🏠 Mon Foyer</h2>
            <p>Bonjour ${name},</p>
            <p>Vous avez été invité(e) à rejoindre l'espace famille <strong>Mon Foyer</strong>.</p>
            <p style="margin:24px 0">
              <a href="${inviteUrl}" style="background:#378ADD;color:#fff;padding:12px 24px;border-radius:8px;text-decoration:none;font-weight:600">
                Créer mon compte →
              </a>
            </p>
            <p style="color:#888;font-size:13px">Ce lien est valable 7 jours.</p>
          </div>
        `,
            }),
        });
    }
    catch (err) {
        console.error('[mailer] fetch error:', err);
        return false;
    }
    if (!res.ok) {
        const body = await res.text().catch(() => '(no body)');
        console.error(`[mailer] Resend API error ${res.status}: ${body}`);
        return false;
    }
    console.log(`[mailer] Invite email sent to ${to}`);
    return true;
}
